decentraland/unity-explorer
# fix: unreliable 'playing' video event

### WHY
Currently we are propagating the `MediaPlayerController` playing states that we read from AVPro directly into the `VideoEvent` "result" component.

The problem is that **AVPro "states" are actually flags**, for example while PLAYING/PAUSED the media player can also be SEEKING and/or BUFFERING.

Additionally I've encountered situations in which the player is PLAYING + SEEKING + BUFFERING but the video is actually frozen... so the `VS_PLAYING` video event state shouldn't be used in that case (`VS_SEEKING` or `VS_BUFFERING` should).

Issue: https://github.com/decentraland/creator-hub/issues/241

### WHAT

* Corrected AVPro flags detection logic to populate the `VideoEvent` component, mainly based on PLAYING vs PAUSED
* While PLAYING-BUFFERING/SEEKING, added an extra **video time check** to deal with the case of the video freezing while AVPro says it's playing, making the video events a lot more reliable for synching
* Added the usage of `VS_ERROR` when the media player has been frozen for 10 seconds or more (can be tweaked in the future) so that scenes can react to that scenario as well

### QA TEST INSTRUCTIONS

The problem is not super easy to reproduce, you have to click the cube (that makes the video start) several times until you see it happening.

**The problem:** sometimes when the video is still "loading" you see the video screen in black but the video event (shown in the top UI) says "PLAYING".

1. Download [this test scene](https://github.com/user-attachments/files/17431289/video-event-seek-test.zip) and uncompress it somewhere
2. Enter the scene root folder and run `npm i` and then `npm run start -- --explorer-alpha`
3. After the scene loads confirm the problem in the scene as shown in the video:

https://github.com/user-attachments/assets/356afa94-97d4-4bde-b6e5-0e8b0a05b5de

4. Close the Explorer that auto-opened. Leave the scene running in the console/terminal.
5. Download the build from this PR and connect it to the running scene using [the console command](https://github.com/decentraland/unity-explorer/wiki/How-to-connect-to-a-local-scene-(Unity-Editor---Custom-Build---Latest-Released-Build)#connecting-a-custom-build-to-the-scene) according to your OS
6. After the scene loads confirm that when you get the black screen before playing, the video event displayed is either BUFFERING or SEEKING (any of those is OK). Video:

https://github.com/user-attachments/assets/4b14a182-de3b-47a6-9f18-dec39c2d8f45
